File: bst.py
# // Time Complexity :O(height)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No

# we implemented two types- recursive and iterative solution. here we check the val we want to find is greater than the root or lesser than the root and move accordingly. 
class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        # recursive solution O(n) O(h)
        # if(root.val>p.val and root.val>q.val):
        #     return self.lowestCommonAncestor(root.left,p,q)
        # elif(root.val<p.val and root.val<q.val):
        #     return self.lowestCommonAncestor(root.right,p,q)
        # else:
        #     return root

        # iterative solution O(h) O(1)
        while(True):
            if(root.val>p.val and root.val>q.val):
                root=root.left
            elif(root.val<p.val and root.val<q.val):
                root=root.right
            else:
                return root

File: 235_Lowest_Common_Ancestor_of_a_Binary_Search_Tree/bst.cpp
#include<bits/stdc++.h>
using namespace std;

/*
Approach: BST property: If at root p and q both are less than root then the lca exists in left subtree. If both are greater then in right subtree. If root = p or q then root is the lca. If p>root but q<root and vice versa then also root is the lca
Time: O(h)
Space: O(h) .. recursive stack
Works on leetcode: Yes
*/

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
public:
    TreeNode* lca;
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        lca = nullptr;
        preorder(root, p, q);
        return lca;
    }

    void preorder(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == nullptr || lca != nullptr) {
            return;
        }

        if(root == p || root == q) {
            lca = root;
            return;
        }

        if((root->val>p->val && root->val<q->val) || (root->val<p->val && root->val>q->val)) {
            lca = root;
            return;
        }
        if(root->val>p->val) {
            preorder(root->left, p, q);
        }
        else {
            preorder(root->right, p, q);
        }
    }
};